### ISSUE TITLE
Execution not stopping and not reporting error on += operator

### ISSUE THREAD

**Issue by dvojtise** [gemoc/ale-lang#53]
The issue https://github.com/gemoc/ale-lang/issues/38 is not completely fixed

There are still some situations where the engine does not stop and does not report any error.

on the following example:

the typechecker does not report any error on the code and on execution, the program continues without any error message.

![image](https://user-images.githubusercontent.com/661468/63250279-8e86d100-c26b-11e9-87f1-5244cd0837a5.png)

**Comment by dvojtise** [comment]
lost several hours due to this one again -> increasing priority

**Comment by echebbi** [comment]
Guess I found the culprit!

https://github.com/gemoc/ale-lang/blob/dc514c8fd46d27b89124dbf43bbcb2424fbbb310/plugins/org.eclipse.emf.ecoretools.ale.core/src/org/eclipse/emf/ecoretools/ale/core/interpreter/MethodEvaluator.java#L204-L214

It seems that `-=` has the same problem. I'll try to fix the issue this afternoon.

**Comment by dvojtise** [comment]
The fix is probably to throw a CriticalFailure exception as in https://github.com/gemoc/ale-lang/commit/fc6fd8d386f1f44ebab18dff36da5b87fb1de962 in the places you've identified ...


As a general rule, it would probably be better to throw this exception in every place where there is a TODO in the code ... so it would be easier to to find them when we use an unsupported /not implemented method

**Comment by echebbi** [comment]
This is indeed the easiest fix I've found. It produces the following output when executing your example:

```
Run helloworld.dsl
------------
start
hws = 
aVar = org.eclipse.emf.ecore.impl.DynamicEObjectImpl@7924de20 (eClass: org.eclipse.emf.ecore.impl.EClassImpl@49bda371 (name: HelloWorld) (instanceClassName: null) (abstract: false, interface: false))
hws size = 1

[AQL eval fail] At line 17 in C:\Progs\IDE\eclipse-dsl-2019-06-R-win32-x86_64\runtime-EclipseApplication\helloworld\model\helloworld.ale :
Diagnostic ERROR source=interpreter code=0 Variable 'self' is not a list, cannot insert org.eclipse.emf.ecore.impl.DynamicEObjectImpl@7435415a (eClass: org.eclipse.emf.ecore.impl.EClassImpl@49bda371 (name: HelloWorld) (instanceClassName: null) (abstract: false, interface: false)) data=[org.eclipse.acceleo.query.ast.impl.CallImpl@3ff43f26 (serviceName: create, type: CALLORAPPLY)]
Variable 'self' is not a list, cannot insert org.eclipse.emf.ecore.impl.DynamicEObjectImpl@7435415a (eClass: org.eclipse.emf.ecore.impl.EClassImpl@49bda371 (name: HelloWorld) (instanceClassName: null) (abstract: false, interface: false))
```
Would it suit you?

**Comment by dvojtise** [comment]
yes for the runtime part which would be much better than nothing, and would allow to decrease the priority :smile: 

The complete fix would also do something on the static typechecker and report the error in the editor too.

**Comment by echebbi** [comment]
Should be fixed by #69
